### ISSUE TITLE
BuilderDeltaConverter leaks JDT JavaModelCache.childrenCache 

### ISSUE THREAD

**Issue by sergio-otero** [eclipse/xtext-eclipse#986]
I've been making performance optimizations on a pretty big project with almost 10.000 source files and after solving XText related problems (for example limiting scope), i'm stuck with an OutOfMemory from JDT itself when the generated java files from my DSL are compiled.

I've been narrowing down the problem and i've found 2 issues:

* OutOfMemory when the files appear due to an interaction with XText BuilderDeltaConverter 
* OutOfMemory when doing a full build because of JDT batch compilation of 2.000 files: this can be solved with -DmaxCompiledUnitsAtOnce=200

The first one only happens with XText plugins installed, even when the project is a simple java project with no xtext nature

Steps to reproduce:

* Open an Eclipse 2018-02 without XText Plugins. Use "-DmaxCompiledUnitsAtOnce=200" in eclipse.ini to avoid the other JDT problem
* Create a simple Java Project and paste the class GenCode below
* Execute GenCode to generate 10.000 files with 2.000 variables each
* Connect JVisualVM to Eclipse with VisualGC
* Refresh the project to let eclipse index the new files
* Look at OldGen occupation growing to 800Mb but going down to 400Mb after finishing (it may need a "perform gc" from JVisualVM)
* Then repeat the same experiment but with XText Plugins installed in Eclipse (happens at least in Eclipse Neon and 2018-12)
* In this case, memory goes up to several Gb and if it has enough to finish, it does not go down less than 2.4Gb. It happens event wild autobuild off
* Doing a heapdump shows that almost all memory is due to org.eclipse.jdt.internal.core.SourceFieldElementInfo and it's root path points to JavaModelManager->JavaModelCache->childrenCache (see attached jdt-dump.png)
* With the help of some breakpoints, i've seen that XText BuilderDeltaConverter somehow triggers that cache been filled when it's designed to hold only information about opened files. See the stacktrace at the bottom

My DSL does not need to link java types like xtend, but it doesn't seem posible to disable BuilderDeltaConverter looking at all the generated java code of some projects

Issue #70 is related, but in my case the java code is inside an Xtext project, so if BuilderDeltaConverter is changed to avoid looking at projects without XText Nature, it will still affect me.

I have a proposed fix in org.eclipse.xtext.common.types.ui.notification.DeltaConverter for convertNewTypes and convertChangedCompilationUnit

```
	/**
	 * @since 2.5
	 */
	protected void convertChangedCompilationUnit(IJavaElementDelta delta, List<IResourceDescription.Delta> result) {
		if ((delta.getFlags() & IJavaElementDelta.F_PRIMARY_WORKING_COPY) != 0) {
			return;
		}
		TypeNames typeNames = getTypeNamesFromIndex(delta);
		
		ICompilationUnit compilationUnit = (ICompilationUnit) delta.getElement();
		try {
			// Use new function to loop over child types
			doForeachCompilationType(compilationUnit, type -> {
				if (isDerived(type))
					return;
				String typeName = type.getFullyQualifiedName();
				URI topLevelURI = uriHelper.createResourceURIForFQN(typeName);
				convertChangedTypeAndChildren(result, typeNames, type, topLevelURI);
			});
				
			convertRemovedTypes(typeNames, result);
		} catch (JavaModelException e) {
			if (LOGGER.isDebugEnabled()) {
				LOGGER.debug(e, e);
			}
		}
	}

	/**
	 * @since 2.5
	 */
	protected void convertNewTypes(ICompilationUnit compilationUnit, List<IResourceDescription.Delta> result) {
		try {
			// Use new function to loop over child types
			doForeachCompilationType(compilationUnit, type -> {
				if (isDerived(type))
					return;
				String typeName = type.getFullyQualifiedName();
				URI topLevelURI = uriHelper.createResourceURIForFQN(typeName);
				convertNewTypeAndChildren(topLevelURI, type, result);
			});
		} catch (JavaModelException e) {
			if (LOGGER.isDebugEnabled())
				LOGGER.debug(e, e);
		}
	}

	private void doForeachCompilationType(ICompilationUnit compilationUnit, Consumer<IType> consumer) throws JavaModelException {
		// Check if unit was open
		IOpenable openableUnit = compilationUnit.getOpenable();
		boolean wasOpen = openableUnit.isOpen();

		try {
			for (IType type : compilationUnit.getTypes()) {
				consumer.accept(type);
			}
		} finally {
			// Close unit if it was not opened
			boolean isOpen = openableUnit.isOpen();
			if (!wasOpen && isOpen) {
				openableUnit.close();
			}
		}
	}
```

```
protected void convertChangedCompilationUnit(IJavaElementDelta delta, List<IResourceDescription.Delta> result) {
		...
		// Check if unit was open
		IOpenable openableUnit = compilationUnit.getOpenable();
		boolean wasOpen = openableUnit.isOpen();
		try {
			...
		} finally {
			// Close unit if it was not opened
			boolean isOpen = openableUnit.isOpen();
			if (!wasOpen && isOpen) {
				try {
					openableUnit.close();
				} catch (JavaModelException e) {
					if (LOGGER.isDebugEnabled()) {
						LOGGER.debug(e, e);
					}
				}
			}
		}
```

```
JavaModelCache

	/**
	 * Cache of open children of openable Java Model Java elements
	 */
	protected Map childrenCache;
```

```
package test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;

public class GenCode {
	private static final int NUM_FILES = 10000;
	private static final int NUM_VARS = 2000;
	
	public static void main(String[] args) throws IOException {
		Path dir = Paths.get("src").resolve("gen");
		
		Files.createDirectories(dir);
		
		String pre = "package gen;\n\n";
	
		StringBuffer sb = new StringBuffer();
		   
		for (int i = 0; i < NUM_FILES; i++) {
			// Class
			sb.append(pre);
			sb.append("public class TEST_").append(i).append(" {\n\n");

			// Vars					
			for (int j = 0; j < NUM_VARS; j++) {
				sb.append("	private String var_").append(j).append(";\n");
			}
			
			// Method to avoid unused variable warning
			sb.append("	public void test() {\n");
			for (int j = 0; j < NUM_VARS; j++) {
				sb.append("		System.out.println(var_").append(j).append(");\n");
			}
			sb.append(" }\n");
			sb.append("}");
			
			//Write class
			Files.write(dir.resolve("TEST_"+i+".java"), sb.toString().getBytes(), StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
			
			sb.setLength(0);
			
			if (i % 1000 == 0) {
				System.out.println(i);
			}
		}
	}
}
```

```
Thread [Worker-17: refresh] (Suspended (breakpoint at line 252 in JavaModelCache))	
	owns: JavaModelManager  (id=65)	
	JavaModelCache.putInfo(IJavaElement, Object) line: 252	
	JavaModelManager.putInfos(IJavaElement, Object, boolean, Map<IJavaElement,Object>) line: 4116	
	CompilationUnit(JavaElement).openWhenClosed(Object, boolean, IProgressMonitor) line: 610	
	CompilationUnit(JavaElement).getElementInfo(IProgressMonitor) line: 326	
	CompilationUnit(JavaElement).getElementInfo() line: 312	
	CompilationUnit(JavaElement).getChildren() line: 267	
	CompilationUnit(JavaElement).getChildrenOfType(int) line: 281	
	CompilationUnit.getTypes() line: 972	
	BuilderDeltaConverter(DeltaConverter).convertNewTypes(ICompilationUnit, List<Delta>) line: 257	
	BuilderDeltaConverter(DeltaConverter).convertAddedPackageFragment(IJavaElementDelta, List<Delta>) line: 497	
	BuilderDeltaConverter(DeltaConverter).convertPackageFragment(IJavaElementDelta, List<Delta>) line: 484	
	BuilderDeltaConverter(DeltaConverter).convertCompilationUnits(IJavaElementDelta, List<Delta>) line: 474	
	BuilderDeltaConverter(DeltaConverter).convertCompilationUnits(IJavaElementDelta, List<Delta>) line: 471	
	BuilderDeltaConverter(DeltaConverter).convertCompilationUnits(IJavaElementDelta, List<Delta>) line: 471	
	BuilderDeltaConverter(DeltaConverter).convertCompilationUnits(IJavaElementDelta, List<Delta>) line: 471	
	BuilderDeltaConverter(DeltaConverter).convert(IJavaElementDelta) line: 148	
	JavaChangeQueueFiller.elementChanged(ElementChangedEvent) line: 38	
	DeltaProcessor$3.run() line: 1739	
	SafeRunner.run(ISafeRunnable) line: 45	
	DeltaProcessor.notifyListeners(IJavaElementDelta, int, IElementChangedListener[], int[], int) line: 1727	
	DeltaProcessor.firePostChangeDelta(IJavaElementDelta, IElementChangedListener[], int[], int) line: 1560	
	DeltaProcessor.fire(IJavaElementDelta, int) line: 1536	
	DeltaProcessor.notifyAndFire(IJavaElementDelta) line: 2257	
	DeltaProcessor.resourceChanged(IResourceChangeEvent) line: 2147	
	DeltaProcessingState.resourceChanged(IResourceChangeEvent) line: 477	
	NotificationManager$1.run() line: 300	
	SafeRunner.run(ISafeRunnable) line: 45	
	NotificationManager.notify(ResourceChangeListenerList$ListenerEntry[], ResourceChangeEvent, boolean) line: 290	
	NotificationManager.broadcastChanges(ElementTree, ResourceChangeEvent, boolean) line: 153	
	Workspace.broadcastPostChange() line: 379	
	Workspace.endOperation(ISchedulingRule, boolean) line: 1502	
	ResourceMgmtActionProvider$1$1(InternalWorkspaceJob).run(IProgressMonitor) line: 49	
	Worker.run() line: 63
![jdt-dump](https://user-images.githubusercontent.com/4890236/53077086-884cad80-34f1-11e9-8281-6aa09ee192ba.png)

```

**Comment by szarnekow** [comment]
Hey sergio, from my understanding the fix is to close the compilation units. Could you please put your findings into a pull request?

**Comment by cdietrich** [comment]
@szarnekow is this one of the build things on your list?

**Comment by cdietrich** [comment]
Fixed in 2.18
